### PR TITLE
[Multi-User Part 5]: Add a production Docker file and use a gunicorn configuration with it

### PR DESCRIPTION
**Pull request by sabaimran**
# Incoming
- Add a productionized setup for the Khoj server using `gunicorn` with multiple workers for handling requests
- Add a new Dockerfile meant for production config at `ghcr.io/khoj-ai/khoj:prod`; the existing Docker config should remain the same

### CODE REVIEW

File: src/khoj/utils/cli.py [hunk 1]
```diff
@@ -42,7 +45,9 @@ def cli(args=None):
         help="Run Khoj in anonymous mode. This does not require any login for connecting users.",
     )
 
-    args = parser.parse_args(args)
+    args, remaining_args = parser.parse_known_args(args)
+
+    logger.debug(f"Ignoring unknown args: {remaining_args}")
```
Comment (debanjum) [inline src/khoj/utils/cli.py:50]: Nit: Log message update for more context
```suggestion
    logger.debug(f"Ignoring unknown commandline args: {remaining_args}")
```

File: prod.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+# Install System Dependencies
+RUN apt update -y && apt -y install python3-pip git
+
+WORKDIR /app
+
+# Install Application
+COPY pyproject.toml .
+COPY README.md .
+RUN sed -i 's/dynamic = \["version"\]/version = "0.0.0"/' pyproject.toml && \
+    TMPDIR=/home/cache/ pip install --cache-dir=/home/cache/ -e .
```
Comment (debanjum) [inline prod.Dockerfile:15]: Why is the tmpdir required to be changed to this location?

File: prod.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+# Copy Source Code
+COPY . .
+
+RUN apt install vim -y
```
Comment (debanjum) [inline prod.Dockerfile:20]: Is vim install required here?

File: .github/workflows/dockerize_production.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      - "*"
+    branches:
+      - master
+    paths:
+      - src/khoj/**
+      - config/**
+      - pyproject.toml
+      - Dockerfile
+      - docker-compose.yml
+      - .github/workflows/dockerize.yml
```
Comment (debanjum) [inline .github/workflows/dockerize_production.yml:16]:
```suggestion
      - .github/workflows/dockerize_production.yml
```

File: .github/workflows/dockerize_production.yml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    tags:
+      - "*"
+    branches:
+      - master
+    paths:
+      - src/khoj/**
+      - config/**
+      - pyproject.toml
+      - Dockerfile
+      - docker-compose.yml
```
Comment (debanjum) [inline .github/workflows/dockerize_production.yml:15]: Should changes to the docker-compose trigger a new docker image build?

File: .github/workflows/dockerize_production.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  push:
+    tags:
+      - "*"
+    branches:
+      - master
+    paths:
+      - src/khoj/**
+      - config/**
+      - pyproject.toml
+      - Dockerfile
```
Comment (debanjum) [inline .github/workflows/dockerize_production.yml:14]:
```suggestion
      - prod.Dockerfile
```

**Review by debanjum — APPROVED** [review]
Left some comments but looks good otherwise

File: .github/workflows/dockerize_production.yml (hunk 5 above)
Comment (sabaimran) [reply .github/workflows/dockerize_production.yml:15]: No, the `docker-compose` is just for use outside of the application. I'll update this.

File: prod.Dockerfile (hunk 3 above)
Comment (sabaimran) [reply prod.Dockerfile:20]: Always useful for debugging, I figured we may as well pre-install it to make it easier to maintain.

**Pull request merged**